plugin/src/main/java/com/google/errorprone/xplat/checker/J2objcMethodName.java

```diff
@@ -0,0 +1,326 @@
+ */
+@BugPattern(
+    name = "J2ObjCMethodName",
+    summary = "Warns against long J2ObjC translated methods.",
+    explanation =
+        "J2ObjC translates Java methods in a specific format that can lead to unreasonably long"
+            + " Objective-C functions when the Java method has many parameters and/or uses long"
+            + " types. This checker calls out problematic methods and offers a fix.",
+    severity = WARNING)
+public class J2objcMethodName extends BugChecker implements MethodTreeMatcher,
+    ClassTreeMatcher, CompilationUnitTreeMatcher {
+
+  private final Map<String, String> foundObjcClassNames = new HashMap<>();
+
+  private final Map<String, Integer> usedMethodNames = new HashMap<>();
+
+  private String packageAnnotation;
+
+  private int methodNameLength = 300;
+
+  private static final Matcher<MethodTree> MATCHER =
+      Matchers.allOf(
+          Matchers.isStatic(),
+          Matchers.methodHasVisibility(Visibility.PUBLIC)
+      );
+
+  private static final Matcher<Tree> OBJC_NAME_MATCHER =
+      Matchers.hasAnnotation("com.google.j2objc.annotations.ObjectiveCName");
+
+  private Description genDescription(String message, MethodTree tree) {
+    return buildDescription(tree)
+        .setMessage(message)
+        .addFix(SuggestedFix.builder()
+            .addImport("com.google.j2objc.annotations.ObjectiveCName")
+            .prefixWith(tree, String.format("@ObjectiveCName(\"%s\")\n",
+                getMethodName(tree.getName().toString())))
+            .build())
+        .build();
+  }
+
+  private String classToCamelCase(String enclosingClass, String outermostClass, Symbol symbol) {
+    StringBuilder newName = new StringBuilder();
+
+    if (foundObjcClassNames.containsKey(symbol.outermostClass().getSimpleName().toString())) {
+      newName.append(foundObjcClassNames.get(symbol.outermostClass().getSimpleName().toString()));
+    } else if (packageAnnotation != null) {
+      newName.append(packageAnnotation);
+      newName.append(symbol.outermostClass().getSimpleName());
+    } else {
+      Iterable<String> packageParts = Splitter.on('.').split(outermostClass);
+
+      for (String part : packageParts) {
+        newName.append(part.substring(0, 1).toUpperCase());
+        newName.append(part.substring(1));
+      }
+    }
+
+    if (outermostClass.length() < enclosingClass.length()) {
+      Iterable<String> remainingParts = Splitter.on('.')
+          .split(enclosingClass.substring(outermostClass.length() + 1));
+
+      for (String part : remainingParts) {
+        newName.append("_");
+
+        if (foundObjcClassNames.containsKey(part)) {
+          newName = new StringBuilder(foundObjcClassNames.get(part));
+        } else {
+          newName.append(part.substring(0, 1).toUpperCase());
+          newName.append(part.substring(1));
+        }
+      }
+    }
+    return newName.toString();
+  }
+
+  private String typeToCamelCase(String enclosingClass, String outermostClass) {
```

Comment:
Does this work with existing types and methods already annotated with `@ObjectiveCName`?

Reply:
typeToCamelCase does not work on classes in the current file, which is the job of classToCamelCase. I have updated the logic to use classToCamelCase if a class inside the source is used as a parameter.